File: Beecrowd_Parte2/HighFive_2539.java
import java.util.Scanner;

public class Main {

    static class FenwickTree {
        int[] tree;
        int size;

        public FenwickTree(int size) {
            this.size = size;
            tree = new int[size + 1];
        }

        public void update(int index, int delta) {
            while (index <= size) {
                tree[index] += delta;
                index += index & -index;
            }
        }

        public int query(int index) {
            int sum = 0;
            while (index > 0) {
                sum += tree[index];
                index -= index & -index;
            }
            return sum;
        }

        public int queryRange(int left, int right) {
            return query(right) - query(left - 1);
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        while (sc.hasNext()) {
            int N = sc.nextInt();
            int[] athletes = new int[N];
            for (int i = 0; i < N; i++) {
                athletes[i] = sc.nextInt();
            }

            FenwickTree fenwickTree = new FenwickTree(N);
            for (int i = 1; i <= N; i++) {
                fenwickTree.update(i, 1);
            }

            long totalHighFives = 0;
            for (int i = 0; i < N; i++) {
                int athleteNumber = athletes[i];
                fenwickTree.update(athleteNumber, -1);
                totalHighFives += fenwickTree.queryRange(athleteNumber + 1, N);
            }

            System.out.println(totalHighFives);
        }

        sc.close();
    }
}

File: Beecrowd_Parte2/HighFive_2539.c
#include <stdio.h>
#include <stdlib.h>

#define MAX_N 100005

typedef struct FenwickTree {
    int size;
    int tree[MAX_N];
} FenwickTree;

void fenwickTree_init(FenwickTree *ft, int size) {
    ft->size = size;
    for (int i = 1; i <= size; i++) {
        ft->tree[i] = 0;
    }
}

void fenwickTree_update(FenwickTree *ft, int index, int delta) {
    while (index <= ft->size) {
        ft->tree[index] += delta;
        index += index & -index;
    }
}

int fenwickTree_query(FenwickTree *ft, int index) {
    int sum = 0;
    while (index > 0) {
        sum += ft->tree[index];
        index -= index & -index;
    }
    return sum;
}

int main() {
    int N;
    while (scanf("%d", &N) != EOF) {
        int athletes[N + 1];
        for (int i = 1; i <= N; i++) {
            scanf("%d", &athletes[i]);
        }

        FenwickTree fenwickTree;
        fenwickTree_init(&fenwickTree, N);

        long long totalHighFives = 0;

        for (int i = 1; i <= N; i++) {
            int athleteNumber = athletes[i];
            totalHighFives += fenwickTree_query(&fenwickTree, N) - fenwickTree_query(&fenwickTree, athleteNumber);
            fenwickTree_update(&fenwickTree, athleteNumber, 1);
        }

        printf("%lld\n", totalHighFives);
    }

    return 0;
}
